core/include/traccc/fitting/kalman_filter/kalman_actor.hpp

```diff
@@ -33,31 +34,50 @@ struct kalman_actor : detray::actor {
         state(vector_t<track_state_type>&& track_states)
             : m_track_states(std::move(track_states)) {
             m_it = m_track_states.begin();
+            m_it_rev = m_track_states.rbegin();
         }
 
         /// Constructor with the vector of track states
         TRACCC_HOST_DEVICE
         state(const vector_t<track_state_type>& track_states)
             : m_track_states(track_states) {
             m_it = m_track_states.begin();
+            m_it_rev = m_track_states.rbegin();
         }
 
         /// @return the reference of track state pointed by the iterator
         TRACCC_HOST_DEVICE
-        track_state_type& operator()() { return *m_it; }
+        track_state_type& operator()() {
+            if (!backward_mode) {
+                return *m_it;
+            } else {
+                return *m_it_rev;
+            }
+        }
 
         /// Reset the iterator
         TRACCC_HOST_DEVICE
-        void reset() { m_it = m_track_states.begin(); }
+        void reset() {
+            m_it = m_track_states.begin();
+            m_it_rev = m_track_states.rbegin();
+        }
 
         /// Advance the iterator
         TRACCC_HOST_DEVICE
-        void next() { m_it++; }
+        void next() {
+            if (!backward_mode) {
+                m_it++;
+            } else {
+                m_it_rev++;
+            }
+        }
 
         /// @return true if the iterator reaches the end of vector
         TRACCC_HOST_DEVICE
         bool is_complete() const {
-            if (m_it == m_track_states.end()) {
+            if (!backward_mode && m_it == m_track_states.end()) {
+                return true;
+            } else if (backward_mode && m_it_rev == m_track_states.rend()) {
```

Comment:
@krasznaa 

Can you figure out why the following error appears (only for CUDA):
```
/mnt/nvme0n1/byeo/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/kalman_actor.hpp(80): error: no operator "==" matches these operands
            operand types are: const vecmem::details::reverse_iterator<traccc::track_state<traccc::default_algebra> *> == vecmem::details::reverse_iterator<const traccc::track_state<traccc::default_algebra> *>
              } else if (backward_mode && m_it_rev == m_track_states.rend()) {
                                                   ^
/mnt/nvme0n1/byeo/projects/traccc/traccc/core/include/traccc/edm/measurement.hpp(79): note #3326-D: function "traccc::operator==" does not match because argument #1 does not match parameter
  inline bool operator==(const measurement& lhs, const measurement& rhs) {
              ^
/mnt/nvme0n1/byeo/projects/traccc/traccc_build/_deps/vecmem-src/core/include/vecmem/containers/impl/reverse_iterator.ipp(161): note #3327-D: candidate function template "vecmem::details::operator==" failed deduction
  __attribute__((host)) __attribute__((device)) bool operator==(const reverse_iterator<T>& itr1,
                                                     ^
/mnt/nvme0n1/byeo/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/kalman_actor.hpp(80): note #3328-D: built-in operator==(<promoted arithmetic>, <promoted arithmetic>) does not match because argument #1 does not match parameter
              } else if (backward_mode && m_it_rev == m_track_states.rend()) {
                                                   ^
/mnt/nvme0n1/byeo/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/kalman_actor.hpp(80): note #3328-D: built-in operator==(<nullptr>, <nullptr>) does not match because argument #1 does not match parameter
              } else if (backward_mode && m_it_rev == m_track_states.rend()) {
                                                   ^
          detected during:
```

Reply:
It's because the right hand side is const and the left hand side is not. I think this will require a fix in vecmem.

Reply:
OK I don't get how `rend` has const type

Reply:
`m_track_states` is const, and calling `rend` on a const value returns a const iterator. `rend` on a non-const value gives a non-const iterator, and `crend` always gives a const iterator.

Reply:
By the way, you can also set `m_it_rev` to be a `const_reverse_iterator`.

Reply:
It compiles after removing the constness of the `is_complete` function

I think the const function tries to call the `const_reverse_iterator` which causes the problem
```c++
    /// Return a reverse iterator pointing at the beginning of the vector
    VECMEM_HOST_AND_DEVICE
    reverse_iterator rend();
    /// Return a constant reverse iterator pointing at the beginning of the
    /// vector
    VECMEM_HOST_AND_DEVICE
    const_reverse_iterator rend() const;
```

Reply:
Ah our comments crossed - I will use the const iterator

Reply:
That's intentional and matches the way [`std::vector::end`](https://en.cppreference.com/w/cpp/container/vector/end) works.

Indeed you can either:

1. Remove the const qualifier on `is_complete` which makes `m_track_states` non-const and thus makes the iterator non-const; or
2. Turn `m_it_rev` from a `reverse_iterator` to a `const_reverse_iterator`; or
3. Modify vecmem to allow you to compare const and non-const iterators.

Reply:
> EDIT: Maybe not - the const iterator conflict with the function of `operator()`..

What do you mean? That sounds like a problem. :thinking:

Reply:
Fix in https://github.com/acts-project/vecmem/pull/307.